### PR TITLE
Dynamic Hub Work Part 1

### PR DESCRIPTION
**Pull request by mikaelm12**
First part of the dynamic hub work
issue: https://github.com/aspnet/SignalR/issues/554

### CODE REVIEW

File: src/Microsoft.AspNetCore.SignalR/Internal/HubReflectionHelper.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@ namespace Microsoft.AspNetCore.SignalR.Internal
 {
     public static class HubReflectionHelper
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.SignalR/Internal/HubReflectionHelper.cs]: Tbh not super sure what to do in this file.

Comment (mikaelm12) [inline src/Microsoft.AspNetCore.SignalR/Internal/HubReflectionHelper.cs]: Tests Pass though 😄

File: src/Microsoft.AspNetCore.SignalR/IHubActivator.cs [hunk 2]
```diff
@@ -5,7 +5,7 @@
 
 namespace Microsoft.AspNetCore.SignalR
 {
-    public interface IHubActivator<THub, TClient> where THub : Hub<TClient>
+    public interface IHubActivator<THub, TClient> where THub : Hub
     {
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/IHubActivator.cs]: This shouldn't have TClient

File: src/Microsoft.AspNetCore.SignalR/HubContext.cs [hunk 3]
```diff
@@ -1,14 +1,9 @@
...
 
-using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Threading.Tasks;
-
 namespace Microsoft.AspNetCore.SignalR
-{
-    public class HubContext<THub> : IHubContext<THub>, IHubConnectionContext<IClientProxy>
+{
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/HubContext.cs]: Nit: remove whitespace

File: src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs [hunk 4]
```diff
@@ -20,7 +20,7 @@
 
 namespace Microsoft.AspNetCore.SignalR
 {
-    public class HubEndPoint<THub> : HubEndPoint<THub, IClientProxy> where THub : Hub<IClientProxy>
+    public class HubEndPoint<THub> : HubEndPoint<THub, IClientProxy> where THub : Hub
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs]: The base type should be removed

File: src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs [hunk 5]
```diff
@@ -32,19 +32,19 @@ public class HubEndPoint<THub> : HubEndPoint<THub, IClientProxy> where THub : Hu
         }
     }
 
-    public class HubEndPoint<THub, TClient> : IInvocationBinder where THub : Hub<TClient>
+    public class HubEndPoint<THub, TClient> : IInvocationBinder where THub : Hub
     {
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs]: There should only be HubEndpoint with a single type argument

File: src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs [hunk 6]
```diff
@@ -32,19 +32,19 @@ public class HubEndPoint<THub> : HubEndPoint<THub, IClientProxy> where THub : Hu
...
         private readonly Dictionary<string, HubMethodDescriptor> _methods = new Dictionary<string, HubMethodDescriptor>(StringComparer.OrdinalIgnoreCase);
 
         private readonly HubLifetimeManager<THub> _lifetimeManager;
-        private readonly IHubContext<THub, TClient> _hubContext;
+        private readonly IHubContext<THub> _hubContext;
         private readonly ILogger<HubEndPoint<THub, TClient>> _logger;
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs]: The logger should also have a single type argument.

File: samples/ChatSample/HubWithPresence.cs [hunk 7]
```diff
@@ -14,7 +14,7 @@ public HubWithPresence(IUserTracker<HubWithPresence> userTracker)
         { }
     }
 
-    public class HubWithPresence<TClient> : Hub<TClient>
```
Comment (davidfowl) [inline samples/ChatSample/HubWithPresence.cs]: Remove TClient

File: src/Microsoft.AspNetCore.SignalR/IHubConnectionContext.cs [hunk 8]
```diff
@@ -3,14 +3,14 @@
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/IHubConnectionContext.cs]: Rename the file to `IHubClients.cs`

**Pull request merged**